internal/internal_eager_activity.go

```diff
@@ -0,0 +1,157 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"fmt"
+	"sync"
+
+	"go.temporal.io/api/workflowservice/v1"
+)
+
+// eagerActivityExecutor is a worker-scoped executor for eager activities that
+// are returned from workflow task completion responses.
+type eagerActivityExecutor struct {
+	eagerActivityExecutorOptions
+
+	activityWorker *activityWorker
+	pendingCount   int // Only access under countLock
+	executingCount int // Only access under countLock
```

Comment:
Why is it important to make a distinction between pending and executing?
Doesn't it complicate the code unnecessarily?
Unless I'm missing something, you could replace these 2 ints and the lock with an atomic int.

Reply:
It is not important now that I am keeping reservation and execution close together here. I may combine them.

> and the lock with an atomic int.

Can't really do this as the actual act of reserving is to obtain a value from a channel which needs to be in the same critical section as check of these values against max concurrent and update of these values.

Reply:
> Can't really do this as the actual act of reserving is to obtain a value from a channel which needs to be in the same critical section as check of these values against max concurrent and update of these values.

You can consume the permits from the counter before taking from the channel and return the unused permits when read from channel would block.